### PR TITLE
[inductor] Take absolute value of strides when picking loop order

### LINKED ISSUE #126860: [inductor][cpu]Background_Matting and pytorch_CycleGAN_and_pix2pix AMP multiple thread static/dynamic shape CPP/default wrapper performance regression
### 🐛 Describe the bug

<p>AMP static shape default wrapper</p><table border="1" class="dataframe table">
  <thead>
    <tr style="text-align: right;">
      <th>suite</th>
      <th>name</th>
      <th>thread</th>
      <th>batch_size_new</th>
      <th>speed_up_new</th>
      <th>inductor_new</th>
      <th>eager_new</th>
      <th>compilation_latency_new</th>
      <th>batch_size_old</th>
      <th>speed_up_old</th>
      <th>inductor_old</th>
      <th>eager_old</th>
      <th>compilation_latency_old</th>
      <th>Ratio Speedup(New/old)</th>
      <th>Eager Ratio(old/new)</th>
      <th>Inductor Ratio(old/new)</th>
      <th>Compilation_latency_Ratio(old/new)</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>torchbench</td>
      <td>Background_Matting</td>
      <td>multiple</td>
      <td>1</td>
      <td>1.249089</td>
      <td>0.087523222</td>
      <td>0.10932429384475799</td>
      <td>25.949658</td>
      <td>1</td>
      <td>1.633883</td>
      <td>0.066992007</td>
      <td>0.10945710137318101</td>
      <td>17.852926</td>
      <td>0.76</td>
      <td>1.0</td>
      <td>0.77</td>
      <td>0.69</td>
    </tr>
    <tr>
      <td>torchbench</td>
      <td>pytorch_CycleGAN_and_pix2pix</td>
      <td>multiple</td>
      <td>1</td>
      <td>1.484149</td>
      <td>0.012310231999999999</td>
      <td>0.018270218512567996</td>
      <td>22.915387</td>
      <td>1</td>
      <td>1.97017</td>
      <td>0.009429855999999999</td>
      <td>0.018578419395519997</td>
      <td>13.873788</td>
      <td>0.75</td>
      <td>1.02</td>
      <td>0.77</td>
      <td>0.61</td>
    </tr>
  </tbody>

</table>

<p>AMP static shape cpp wrapper</p><table border="1" class="dataframe table">
  <thead>
    <tr style="text-align: right;">
      <th>suite</th>
      <th>name</th>
      <th>thread</th>
      <th>batch_size_new</th>
      <th>speed_up_new</th>
      <th>inductor_new</th>
      <th>eager_new</th>
      <th>compilation_latency_new</th>
      <th>batch_size_old</th>
      <th>speed_up_old</th>
      <th>inductor_old</th>
      <th>eager_old</th>
      <th>compilation_latency_old</th>
      <th>Ratio Speedup(New/old)</th>
      <th>Eager Ratio(old/new)</th>
      <th>Inductor Ratio(old/new)</th>
      <th>Compilation_latency_Ratio(old/new)</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>torchbench</td>
      <td>Background_Matting</td>
      <td>multiple</td>
      <td>1</td>
      <td>1.259638</td>
      <td>0.08657777900000001</td>
      <td>0.10905666038400201</td>
      <td>26.391435</td>
      <td>1</td>
      <td>1.660248</td>
      <td>0.065993515</td>
      <td>0.10956560129172</td>
      <td>18.38254</td>
      <td>0.76</td>
      <td>1.0</td>
      <td>0.76</td>
      <td>0.7</td>
    </tr>
    <tr>
      <td>torchbench</td>
      <td>pytorch_CycleGAN_and_pix2pix</td>
      <td>multiple</td>
      <td>1</td>
      <td>1.493959</td>
      <td>0.01200208</td>
      <td>0.01793061543472</td>
      <td>23.653602</td>
      <td>1</td>
      <td>2.006529</td>
      <td>0.009302916</td>
      <td>0.018666570738564</td>
      <td>14.677247</td>
      <td>0.74</td>
      <td>1.04</td>
      <td>0.78</td>
      <td>0.62</td>
    </tr>
  </tbody>

</table>

<p>AMP dynamic shape default wrapper</p><table border="1" class="dataframe table">
  <thead>
    <tr style="text-align: right;">
      <th>suite</th>
      <th>name</th>
      <th>thread</th>
      <th>batch_size_new</th>
      <th>speed_up_new</th>
      <th>inductor_new</th>
      <th>eager_new</th>
      <th>compilation_latency_new</th>
      <th>batch_size_old</th>
      <th>speed_up_old</th>
      <th>inductor_old</th>
      <th>eager_old</th>
      <th>compilation_latency_old</th>
      <th>Ratio Speedup(New/old)</th>
      <th>Eager Ratio(old/new)</th>
      <th>Inductor Ratio(old/new)</th>
      <th>Compilation_latency_Ratio(old/new)</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>torchbench</td>
      <td>Background_Matting</td>
      <td>multiple</td>
      <td>1</td>
      <td>1.260694</td>
      <td>0.087103355</td>
      <td>0.10981067702837</td>
      <td>25.825552</td>
      <td>1</td>
      <td>1.639151</td>
      <td>0.06659711900000001</td>
      <td>0.10916273420596902</td>
      <td>17.869419</td>
      <td>0.77</td>
      <td>0.99</td>
      <td>0.76</td>
      <td>0.69</td>
    </tr>
    <tr>
      <td>torchbench</td>
      <td>pytorch_CycleGAN_and_pix2pix</td>
      <td>multiple</td>
      <td>1</td>
      <td>1.503163</td>
      <td>0.012218454</td>
      <td>0.018366327970002</td>
      <td>22.738695</td>
      <td>1</td>
      <td>1.976195</td>
      <td>0.009377097</td>
      <td>0.018530972205914997</td>
      <td>13.877887</td>
      <td>0.76</td>
      <td>1.01</td>
      <td>0.77</td>
      <td>0.61</td>
    </tr>
  </tbody>

</table>

<p>AMP dynamic shape cpp wrapper</p><table border="1" class="dataframe table">
  <thead>
    <tr style="text-align: right;">
      <th>suite</th>
      <th>name</th>
      <th>thread</th>
      <th>batch_size_new</th>
      <th>speed_up_new</th>
      <th>inductor_new</th>
      <th>eager_new</th>
      <th>compilation_latency_new</th>
      <th>batch_size_old</th>
      <th>speed_up_old</th>
      <th>inductor_old</th>
      <th>eager_old</th>
      <th>compilation_latency_old</th>
      <th>Ratio Speedup(New/old)</th>
      <th>Eager Ratio(old/new)</th>
      <th>Inductor Ratio(old/new)</th>
      <th>Compilation_latency_Ratio(old/new)</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>torchbench</td>
      <td>Background_Matting</td>
      <td>multiple</td>
      <td>1</td>
      <td>1.255357</td>
      <td>0.08692128099999999</td>
      <td>0.10911723855231699</td>
      <td>26.434504</td>
      <td>1</td>
      <td>1.661939</td>
      <td>0.06580516900000001</td>
      <td>0.10936417676269101</td>
      <td>18.408933</td>
      <td>0.76</td>
      <td>1.0</td>
      <td>0.76</td>
      <td>0.7</td>
    </tr>
    <tr>
      <td>torchbench</td>
      <td>pytorch_CycleGAN_and_pix2pix</td>
      <td>multiple</td>
      <td>1</td>
      <td>1.445838</td>
      <td>0.01240409</td>
      <td>0.01793430467742</td>
      <td>23.729033</td>
      <td>1</td>
      <td>1.942938</td>
      <td>0.009227029</td>
      <td>0.017927545271201998</td>
      <td>14.852708</td>
      <td>0.74</td>
      <td>1.0</td>
      <td>0.74</td>
      <td>0.63</td>
    </tr>
  </tbody>

</table><p>SW info</p><table border="1" class="dataframe table">
  <thead>
    <tr style="text-align: right;">
      <th>name</th>
      <th>target_branch</th>
      <th>target_commit</th>
      <th>refer_branch</th>
      <th>refer_commit</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>torchbench</td>
      <td>main</td>
      <td>d6015d42</td>
      <td>main</td>
      <td>d6015d42</td>
    </tr>
    <tr>
      <td>torch</td>
      <td>main</td>
      <td>5fb11cda4fe60c1a7b30e6c844f84ce8933ef953</td>
      <td>main</td>
      <td>02093b6c6ae1046368e2500881d0bb5880873386</td>
    </tr>
    <tr>
      <td>torchvision</td>
      <td>main</td>
      <td>0.19.0a0+d23a6e1</td>
      <td>main</td>
      <td>0.19.0a0+d23a6e1</td>
    </tr>
    <tr>
      <td>torchtext</td>
      <td>main</td>
      <td>0.16.0a0+b0ebddc</td>
      <td>main</td>
      <td>0.16.0a0+b0ebddc</td>
    </tr>
    <tr>
      <td>torchaudio</td>
      <td>main</td>
      <td>2.2.0a0+1980f8a</td>
      <td>main</td>
      <td>2.2.0a0+ea437b3</td>
    </tr>
    <tr>
      <td>torchdata</td>
      <td>main</td>
      <td>0.7.1a0+0790338</td>
      <td>main</td>
      <td>0.7.1a0+0790338</td>
    </tr>
    <tr>
      <td>dynamo_benchmarks</td>
      <td>main</td>
      <td>nightly</td>
      <td>main</td>
      <td>nightly</td>
    </tr>
  </tbody>
</table>

</table>

Repro:
[inductor_single_run.sh](https://github.com/chuanqi129/inductor-tools/blob/main/scripts/modelbench/inductor_single_run.sh)
bash inductor_single_run.sh multiple inference performance torchbench model amp first dynamic/static default/cpp
Suspected guilty commit: https://github.com/pytorch/pytorch/commit/f6a00a8032184015358a438d2f7c08ac6a5e36e6
[torchbench-pytorch_CycleGAN_and_pix2pix-inference-amp-dynamic-cpp-multiple-performance-drop_guilty_commit.log](https://github.com/pytorch/pytorch/files/15401405/torchbench-pytorch_CycleGAN_and_pix2pix-inference-amp-dynamic-cpp-multiple-performance-drop_guilty_commit.log)
cc @msaroufim @ezyang @bdhirsh @anijain2305 @chauhang @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @WeizhuoZhang-intel @chuanqi129

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #127425

Fixes #126860

The stride hint is found by comparing the value of the indexing expression
evaluated at `idx` set to all zeros and at `idx[dim] = 1`. This causes a problem
for padded inputs where 0 and 1 are still in the padded region.

In particular, for reflection padding this causes the stride to be negative.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Comment by peterbell10** [comment]
HUD shows improvements to `Background_Matting` and `pytorch_CycleGAN_and_pix2pix` on CUDA, but I don't have the hardware to check on CPU. @jgong5 perhaps someone from intel could run the benchmarks to see if this fixes things?

https://hud.pytorch.org/benchmark/torchbench/inductor_with_cudagraphs?startTime=Mon%2C%2027%20May%202024%2015%3A19%3A18%20GMT&stopTime=Mon%2C%2003%20Jun%202024%2015%3A19%3A18%20GMT&granularity=hour&mode=training&dtype=amp&lBranch=gh/peterbell10/731/orig&lCommit=6ed807f8507c01e647f75b1ce56b519dd552a013&rBranch=main&rCommit=d6e3e89804c4063827ea21ffcd3d865e5fe365d9

**Review by lezcano — APPROVED** [review]
The fix is reasonable and the HUD agrees

**Comment by lezcano** [comment]
Could you also add a regression test?
